### ISSUE TITLE
Inconsistent chunks for deleted and recreated question

### ISSUE THREAD

**Issue by mwest1066** [PrairieLearn/PrairieLearn#3198]
`pl-math221` has question `triple_chain_rule_symbolic_MC_tan`. It was deleted and re-created with the same QID but a different UUID:

```
* 22f85ca 2020-09-21 REDACTED triple_chain_rule_symbolic_MC_tan: edit question.html
* 9920feb 2020-09-21 REDACTED rename question New_2 to triple_chain_rule_symbolic_MC_tan
* cdaba66 2020-09-21 REDACTED New_2: edit info.json
* ee29db4 2020-09-21 REDACTED add question New_2
* ae6adcb 2020-09-21 REDACTED delete question triple_chain_rule_symbolic_MC_tan
* 29eabc5 2020-09-21 REDACTED triple_chain_rule_symbolic_MC_tan: edit info.json
* 5e28477 2020-09-21 REDACTED rename question New_2 to triple_chain_rule_symbolic_MC_tan
* 42ad7b4 2020-09-21 REDACTED add question New_2
```

In the DB we have two question rows (also note the `sync_warnings` about a UUID conflict):

```
-[ RECORD 1 ]----------------------+------------------------------------------------------------------------------------------------------
id                                 | 9047234
qid                                | triple_chain_rule_symbolic_MC_tan
title                              | Triple chain rule MC using sympy tan
course_id                          | 191
deleted_at                         |
uuid                               | b3eed8b3-b6c8-417c-b0f8-62629bd679dd
sync_errors                        |
sync_warnings                      |
-[ RECORD 2 ]----------------------+------------------------------------------------------------------------------------------------------
id                                 | 9047233
qid                                | triple_chain_rule_symbolic_MC_tan
title                              | Triple chain rule MC using sympy tan
course_id                          | 191
deleted_at                         | 2020-09-21 03:01:39.912735+00
uuid                               |
sync_errors                        |
sync_warnings                      | UUID "858178a4-3e3b-462d-b0fd-46e788944e92" is used in other questions: triple_chain_rule_symbolic_MC
```

Corresponding to these are two chunk rows:

```
prairielearn=> select * from chunks where question_id = 9047233;
-[ RECORD 1 ]------+-------------------------------------
id                 | 22799
uuid               | ddce1500-a452-4c76-81ff-597c97af5b8e
type               | question
course_id          | 191
course_instance_id |
assessment_id      |
question_id        | 9047233

prairielearn=> select * from chunks where question_id = 9047234;
-[ RECORD 1 ]------+-------------------------------------
id                 | 22803
uuid               | fc430434-6717-4072-aa45-69ca2b0b1e2e
type               | question
course_id          | 191
course_instance_id |
assessment_id      |
question_id        | 9047234
```

The server `chunk-2` has both of these chunks and the question directory for `/tmp/pl-chunked-courses/course-191/questions/triple_chain_rule_symbolic_MC_tan` is the unpacked version of `fc430434-6717-4072-aa45-69ca2b0b1e2e`, which is correct.

In contrast, server `chunk-instance.prairielearn.org` has `ddce1500-a452-4c76-81ff-597c97af5b8e` unpacked into `/tmp/pl-chunked-courses/course-191/questions/triple_chain_rule_symbolic_MC_tan`. This is:

```
[centos@ip-10-0-3-199 ~]$ cd /tmp/pl-chunked-courses/course-191/questions/triple_chain_rule_symbolic_MC_tan/
[centos@ip-10-0-3-199 triple_chain_rule_symbolic_MC_tan]$ cat info.json
{
    "uuid": "b820b946-da4f-45e5-adac-591e91e7f7d7",
    "title": "New (2)",
    "topic": "Algebra",
    "tags": [
        "mwest",
        "fa17",
        "tpl101",
        "v3"
    ],
    "type": "v3"
}
[centos@ip-10-0-3-199 triple_chain_rule_symbolic_MC_tan]$ cat question.html
<pl-question-panel>
  <p> Consider two numbers $a = {{params.a}}$ and $b = {{params.b}}$.</p>
  <p> What is the sum $c = a + b$?</p>
</pl-question-panel>

<pl-number-input answers-name="c" comparison="sigfig" digits="3" label="$c=$"></pl-number-input>
[centos@ip-10-0-3-199 triple_chain_rule_symbolic_MC_tan]$ cat server.py
import random, copy

def generate(data):

    # Sample two random integers between 5 and 10 (inclusive)
    a = random.randint(5, 10)
    b = random.randint(5, 10)

    # Put these two integers into data['params']
    data['params']['a'] = a
    data['params']['b'] = b

    # Compute the sum of these two integers
    c = a + b

    # Put the sum into data['correct_answers']
    data['correct_answers']['c'] = c
```

However, server `chunk-instance.prairielearn.org` has recently served this question (as `instance_question_id = 99429406`) in student mode:

```
[centos@ip-10-0-3-199 prairielearn_logs]$ grep 99429406 *
prairielearn.log:{"timestamp":"2020-09-26T18:36:52.711Z","ip":"10.0.3.34","forwardedIP":"REDACTED, 10.0.3.34","authn_user_id": REDACTED,"authn_user_uid":"REDACTED","method":"GET","path":"/pl/course_instance/123748/instance_question/99429406/","params":{},"body":{},"response_id":"45cb4af4-f79d-4fda-a297-5c7a4b65ac1f","level":"verbose","message":"request"}
prairielearn.log:{"response_id":"45cb4af4-f79d-4fda-a297-5c7a4b65ac1f","timestamp":"2020-09-26T18:36:52.829Z","ip":"10.0.3.34","forwardedIP":"REDACTED, 10.0.3.34","method":"GET","path":"/","params":{},"body":{},"authn_user_id":798753,"authn_user_uid":"REDACTED","user_id":798753,"user_uid":"REDACTED","course_id":191,"course_short_name":"MATH 221","course_instance_id":123748,"course_instance_short_name":"Fa20","assessment_id":2306886,"assessment_directory":"REDACTED","assessment_instance_id":1829104,"question_id":9047234,"question_directory":"triple_chain_rule_symbolic_MC_tan","instance_question_id":99429406,"level":"verbose","message":"response"}
prairielearn.log:{"response_id":"57688bba-db9e-4f19-919c-152d8baad7de","timestamp":"2020-09-26T18:37:05.293Z","ip":"10.0.0.30","forwardedIP":"REDACTED, 10.0.0.30","method":"GET","path":"/pl/course_instance/123748/instance_question/99429406/","params":{},"body":{},"authn_user_id":null,"authn_user_uid":null,"user_id":null,"user_uid":null,"course_id":null,"course_short_name":null,"course_instance_id":null,"course_instance_short_name":null,"assessment_id":null,"assessment_directory":null,"assessment_instance_id":null,"question_id":null,"question_directory":null,"instance_question_id":null,"level":"verbose","message":"response"}
```

**Comment by nicknytko** [comment]
Following up on this, I think something weird might be happening in `ensureChunk()`. The new chunk `fc430434-6717-4072-aa45-69ca2b0b1e2e.tar.gz` was on disk but for some reason was not unpacked?  Removing the tar from disk and forcing the chunk to re-download and extract seemed to work, but I'd like to understand why it was never extracted in the first place.

**Comment by mwest1066** [comment]
I agree that this seems impossible. If we have `fc430434-6717-4072-aa45-69ca2b0b1e2e.tar.gz` on disk then it looks like we must have executed https://github.com/PrairieLearn/PrairieLearn/blob/b11c990682600c9e3b0b626747ad959eb6e338d5/lib/chunks.js#L707 which can only happen after the unpack succeeds.

Is that your logic as well?

**Comment by mwest1066** [comment]
I'm suspicious that this happened for a question with a complicated edit history (originally a duplicate UUID, then a new question with a recycled QID). I don't see the error path yet, but I suspect that the underlying issue in #3136 is also involved here. It's possible that fixing #3136 will also fix this issue, although I'd like to understand the error path.

**Comment by mwest1066** [comment]
How's this for a theory: before we deployed #3187 the chunk metadata for any given question was being paired with a random other question. This is because `(chunks_arr->>'question_id' IS NULL)` was always TRUE (because we didn't have a `question_id` property) so the only used WHERE conditions in `select_metadata_for_chunks` were the conditions on `chunks.course_id` and `chunks.type`.

This means that we would have matched question `id = 9047234` (which has `q.qid = 'triple_chain_rule_symbolic_MC_tan'`) with _all_ question chunks. We then would have downloaded and unpacked any chunks which hadn't been unpacked already, and all of these would have been unpacked into `questions/triple_chain_rule_symbolic_MC_tan`. It's entirely possible that chunk `ddce1500` was the last one to be unpacked, overwriting `fc430434`.

The one thing I don't understand is why we have so few chunks downloaded:
```
[root@ip-10-0-3-199 chunks]# ll
total 80
-rw-r--r-- 1 root root  721 Sep 24 16:30 0902ada6-c391-447a-965e-0412a7f6ace9.tar.gz
-rw-r--r-- 1 root root  772 Sep 24 16:30 0a741565-21ec-4e78-bba7-bfd9c374a9dd.tar.gz
-rw-r--r-- 1 root root  732 Sep 27 16:29 13b35224-e9d7-4f2d-a521-9c2514f7b65b.tar.gz
-rw-r--r-- 1 root root  678 Sep 24 16:31 21daf5c1-0e03-4a87-ac1e-e543cd41545c-staging.tar.gz
-rw-r--r-- 1 root root 1038 Sep 24 16:30 25adea3f-134e-4dbf-87ab-a24fb799dec1.tar.gz
-rw-r--r-- 1 root root  861 Sep 27 16:29 2efb5257-f729-4fef-855e-fc8b5a3d84fd.tar.gz
-rw-r--r-- 1 root root  752 Sep 24 16:30 404f15b3-5b5f-4e37-81f1-477ac76b845c.tar.gz
-rw-r--r-- 1 root root  681 Sep 24 16:30 4181d6f2-ea66-44fe-8ec5-29fdfd17e9fc.tar.gz
-rw-r--r-- 1 root root  829 Sep 27 16:29 443f0b9b-0c74-4f80-a063-07520d360050.tar.gz
-rw-r--r-- 1 root root  765 Sep 24 16:30 7f8a4e25-fe64-4433-a54b-7d199aa73541.tar.gz
-rw-r--r-- 1 root root 1096 Sep 24 16:30 80526ede-5bcf-41ad-875f-75b941389dea.tar.gz
-rw-r--r-- 1 root root  798 Sep 24 16:30 8f039bcf-651c-472a-a64f-f1869cdf032b.tar.gz
-rw-r--r-- 1 root root  912 Sep 27 16:29 905df83f-0520-451a-b219-7789b0c70def.tar.gz
-rw-r--r-- 1 root root  773 Sep 24 16:30 9718d789-4814-401c-a2eb-37d317761f45.tar.gz
-rw-r--r-- 1 root root  832 Sep 24 16:30 c535df75-f615-4b01-be23-0fd62d02e5dc.tar.gz
-rw-r--r-- 1 root root  785 Sep 27 16:29 c6cd72d4-3bda-4713-a7cc-397686261523.tar.gz
-rw-r--r-- 1 root root  651 Sep 24 16:30 ddce1500-a452-4c76-81ff-597c97af5b8e.tar.gz
-rw-r--r-- 1 root root 1117 Sep 27 16:29 f8e9b517-7a1c-4f10-beb6-b9e5a4afb38a.tar.gz
-rw-r--r-- 1 root root 1077 Sep 27 16:42 fc430434-6717-4072-aa45-69ca2b0b1e2e.tar.gz
-rw-r--r-- 1 root root  740 Sep 24 16:30 ffa51cbc-d529-458e-87a7-70eac917ae8e.tar.gz
```

The course has 73 questions (and 74 chunks) so with this all-to-all matching before #3187 was fixed, I would have thought we'd download all the chunks.

Somehow #3187 was less catastrophic than I think it should have been. Why was the chunk download code working at all before #3187 was fixed? Why was it only this duplicate-QID case that triggered the problem fixed by #3187?

(Note that the reason there are 74 chunks for 73 questions is precisely because of this particular question, which has a deleted question with the same QID.)

**Comment by mwest1066** [comment]
Following up on the issue above that the pre-#3187 version of `select_metadata_for_chunks` would have matched all DB-chunks for any question, this can be verified with the query below (the pre-#3187 query), which returns 74 rows (all chunks) for a single question. However, it does have the correct `question_name` for each chunk, because these are matched using `q.id = chunks.question_id`. This means that a chunk will only ever be unpacked into the correct question directory.

The problem comes with having two questions with the same QID (one of which is deleted). Both of these chunks will have the same `question_name` and the unpack order is random, so we might get the old one. This explains how we could have got `ddce1500` unpacked over `fc430434` in `questions/triple_chain_rule_symbolic_MC_tan`.

Note that this still doesn't explain why we didn't download the chunks for all 73 questions (and unpack them into the correct directories).

```
SELECT
    chunks.*,
    q.qid AS question_name,
    a.tid AS assessment_name,
    ci.short_name AS course_instance_name
FROM
    JSON_ARRAY_ELEMENTS('[{"type":"question", "question_id": 3242}]'::json) AS chunks_arr,
    chunks
    LEFT JOIN assessments AS a ON (a.id = chunks.assessment_id)
    LEFT JOIN course_instances AS ci ON (ci.id = chunks.course_instance_id OR ci.id = a.course_instance_id)
    LEFT JOIN questions AS q ON (q.id = chunks.question_id)
WHERE
    chunks.course_id = 191
    AND chunks.type = (chunks_arr->>'type')::enum_chunk_type
    AND ((chunks_arr->>'courseInstanceId' IS NULL) OR (chunks.course_instance_id = (chunks_arr->>'courseInstanceId')::bigint))
    AND ((chunks_arr->>'assessmentId' IS NULL) OR (chunks.assessment_id = (chunks_arr->>'assessmentId')::bigint))
    AND ((chunks_arr->>'questionId' IS NULL) OR (chunks.question_id = (chunks_arr->>'questionId')::bigint));
```

(query above returns 74 rows, which is all the chunks in the course)

**Comment by mwest1066** [comment]
@nicknytko made the excellent point that the chunks are downloaded within a `eachLimit`: https://github.com/PrairieLearn/PrairieLearn/blob/b11c990682600c9e3b0b626747ad959eb6e338d5/lib/chunks.js#L740 This means that we will stop on the first error, and we in fact did get the error #3185 (which was fixed by #3187). The chunk server also had a partially-synced chunk on disk (like `UUID-staging.tar.gz`) which is consistent with the chunk downloads being interrupted.

Based on this we believe that this issue has in fact already been addressed by #3187.
